starky/src/proof.rs

```diff
@@ -176,22 +279,94 @@ impl<F: RichField + Extendable<D>, const D: usize> StarkOpeningSet<F, D> {
                 .copied()
                 .collect_vec(),
         };
-        FriOpenings {
-            batches: vec![zeta_batch, zeta_next_batch],
+
+        if let Some(ctl_zs_first) = self.ctl_zs_first.as_ref() {
+            debug_assert!(!ctl_zs_first.is_empty());
+            debug_assert!(self.auxiliary_polys.is_some());
+            debug_assert!(self.auxiliary_polys_next.is_some());
+
+            let ctl_first_batch = FriOpeningBatch {
+                values: ctl_zs_first
+                    .iter()
+                    .copied()
+                    .map(F::Extension::from_basefield)
+                    .collect(),
+            };
+
+            FriOpenings {
+                batches: vec![zeta_batch, zeta_next_batch, ctl_first_batch],
+            }
+        } else {
+            FriOpenings {
+                batches: vec![zeta_batch, zeta_next_batch],
+            }
         }
     }
 }
 
+#[derive(Clone, Debug, PartialEq, Eq)]
 pub struct StarkOpeningSetTarget<const D: usize> {
     pub local_values: Vec<ExtensionTarget<D>>,
     pub next_values: Vec<ExtensionTarget<D>>,
     pub auxiliary_polys: Option<Vec<ExtensionTarget<D>>>,
     pub auxiliary_polys_next: Option<Vec<ExtensionTarget<D>>>,
+    pub ctl_zs_first: Option<Vec<Target>>,
     pub quotient_polys: Vec<ExtensionTarget<D>>,
 }
 
 impl<const D: usize> StarkOpeningSetTarget<D> {
-    pub(crate) fn to_fri_openings(&self) -> FriOpeningsTarget<D> {
+    /// Serializes a STARK's opening set.
+    pub(crate) fn to_buffer(&self, buffer: &mut Vec<u8>) -> IoResult<()> {
```

Comment:
Same.

In idiomatic Rust, `to_` methods should do some sort of (possibly expensive) _conversion_, which is not really the case here. This method is instead performing the IO into a buffer. Assuming we have only one way of serializing a STARK, why not call this (and similar methods) "serialize" instead?

Reply:
Alternatively make this into an actual `to_` conversion that takes `Bytes` and returns a `Result<Buffer, _>`?

Reply:
It initially came from the distinction from all the `to_bytes` / `from_bytes` methods, where we needed access to the input/output buffer, hence the function name change to `to_buffer` / `from_buffer`.

Reply:
If we want to update all the serialization method names, I'm not against it *but* this is out of scope of this PR.